lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -535,49 +489,58 @@ scheduleLoads(scf::ForOp forOp, tt::CoarseSchedule &schedule,
     }
   });
   if (loadOpToIndLevelAndUse.empty())
-    return {};
+    return afterPrologue;
 
   // Check which loads are good for pipelining, and assign them
   // memory layouts.
-  llvm::MapVector<Operation *, LoadInfo> loadToInfo =
-      assignMemoryLayouts(loadOpToIndLevelAndUse, axisInfoAnalysis);
+  assignMemoryLayouts(loadOpToIndLevelAndUse, axisInfoAnalysis, loadToInfo);
 
   if (loadToInfo.empty())
-    return {};
+    return afterPrologue;
 
-  // Calculate the stage distance between applicable loads.
-  int maxIndirectionLevel = -1;
-  for (auto [loadOp, dist, use] : loadOpToIndLevelAndUse) {
-    if (loadToInfo.count(loadOp) == 0)
-      continue;
-    maxIndirectionLevel = std::max(maxIndirectionLevel, dist);
-  }
-  unsigned stagesBetweenLoads =
-      ceil<unsigned>(numStages - 2, maxIndirectionLevel + 1);
-
-  tt::CoarseSchedule::Cluster rootUsersCluster = schedule.clusters.newAtFront();
-  // Put the root uses of the loads in the last stage.
-  for (auto &[loadOp, dist, use] : loadOpToIndLevelAndUse) {
-    if (loadToInfo.count(loadOp) == 0)
-      continue;
-    // Non-LoadOp(s) are the root uses of all LoadOp(s) and should be
-    // always present in the opInfo
-    if (!isa<tt::LoadOp>(use)) {
-      schedule.insert(use, numStages - 1, rootUsersCluster);
-      rootUsers.insert(use);
+  // reconstrcut schedule from annotations of (stage, cluster)
+  int maxClusterId = 0, minClusterId = 0;
+  bool hasSchedule = false;
+  for (Operation &op : forOp.getBody()->without_terminator()) {
+    if (op.hasAttr("loop.stage") && op.hasAttr("loop.cluster")) {
+      auto clusterId = cast<IntegerAttr>(op.getAttr("loop.cluster"))
+                           .getValue()
+                           .getSExtValue();
+      LLVM_DEBUG({
+        LDBG("saw cluster " << clusterId);
+        op.dump();
+      });
+      if (!hasSchedule) {
+        minClusterId = clusterId;
+        maxClusterId = clusterId;
+        hasSchedule = true;
+        continue;
+      }
+      minClusterId = (clusterId < minClusterId) ? clusterId : minClusterId;
+      maxClusterId = (clusterId > maxClusterId) ? clusterId : maxClusterId;
     }
   }
-
-  SmallVector<tt::CoarseSchedule::Cluster> loadsClusters;
-  for (int i = 0; i < maxIndirectionLevel + 1; i++) {
-    loadsClusters.push_back(schedule.clusters.newAtBack());
+  assert(hasSchedule);
+  LDBG("minCluster " << minClusterId << " max " << maxClusterId);
+  DenseMap<int, tt::CoarseSchedule::Cluster> clusters;
+  for (int i = minClusterId; i < maxClusterId + 1; i++) {
+    clusters.insert({i, schedule.clusters.newAtBack()});
   }
-  // Assign stages to the loads.
-  for (auto [loadOp, indLevel, _] : loadOpToIndLevelAndUse) {
-    if (loadToInfo.count(loadOp) == 0)
-      continue;
-    int stage = (maxIndirectionLevel - indLevel) * stagesBetweenLoads;
-    schedule.insert(loadOp, stage, loadsClusters[indLevel]);
+  // afterPrologue should be the first cluster after ifOps?
```

Comment:
Sorry the comment is outdated. We need to return afterPrologue since it is needed in scheduleRemaining. The current rule between loopScheduling and SWP is that cluster with clusterId 0 will be afterPrologue (i.e clusterId for prologue in the case of 1D flattened persistent kernel will have negative clusterId from loopScheduling pass). This is an extra communication between the two passes which we should try to remove.